### DISCUSSION TITLE
Access Logging with value from custom header AND optionally response

### DISCUSSION THREAD

**Discussion by ssproessig-thales** [Kludex/uvicorn#1615] [Q&A]
Hi,
I have two questions regarding the access logging

## 1. adding more information from headers
currently, my log looks like this:

```
whatever-api_server-1  | 2022-08-26 14:12:41,963      INFO    10.220.177.148:0 - "PUT /api/v1/me/ping HTTP/1.0" 204
```

I'd like to add some decoded information from a header (namely the user part of `Basic` authentication in the `Authorization` header) so that it looks like

```
whatever-api_server-1  | 2022-08-26 14:12:41,963      INFO    10.220.177.148:0 - api_admin   "PUT /api/v1/me/ping HTTP/1.0" 204
```
from the HTTP header
```
Authorization: Basic YXBpX2FkbWluOng=
```
so at best by invoking a callback that performs base64 decoding and only returns the user part, not the password part. Hence I don't want the raw header to be logged.

Can this be done with uvicorn?

## 2. conditionally log the request / response body
when my API encounters 409 or 422 responses, it would really be helpful to have the request and response body as part of the `DEBUG` log. Currently I have to implement this as per endpoint, which is repetitive and surely is forgotten at some endpoints.

Can this be configured directly in uvicorn?

**Comment by Kludex** [comment]
We should aim to improve the documentation on that, but for now: https://github.com/Kludex/asgi-logger

**Reply by ssproessig-thales** [reply]
thank you very much - I'll check it and get back here to "mark as answer"

**Comment by aranvir** [comment]
Hi, did anything ever come from that? E.g., on how to include header information like `x-forwarded-for` in the access log?

**Reply by aranvir** [reply]
Managed to figure this out after more digging in the issues. One needs to set both `--proxy-headers` and `--forwarded-allow-ips "*"`. For more fine control, provide a list of trusted IPs however, they dont accept wildcards, e.g., something like `123.45.0.*` does not seem to work.
